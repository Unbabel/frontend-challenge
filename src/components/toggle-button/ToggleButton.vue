<script setup>
	import { computed } from "vue";

	const { checked, onChangeCallback } = defineProps({
		checked: {
			type: Boolean,
			required: true,
		},
		onChangeCallback: {
			type: Function,
			required: true,
		},
	});

	const isChecked = computed(() => checked);
</script>

<template>
	<div>
		<label class="switch">
			<input type="checkbox" :checked="isChecked" teste="lol" />
			<span
				class="slider"
				@click="onChangeCallback"
				data-test="toggle-button"
			></span>
		</label>
	</div>
</template>

<style scoped>
	.switch {
		position: relative;
		display: inline-block;
		width: 3rem;
		height: 1.5rem;
	}

	.switch input {
		opacity: 0;
		width: 0;
		height: 0;
	}

	.slider {
		position: absolute;
		cursor: pointer;
		top: 0;
		left: 0;
		right: 0;
		bottom: 0;
		background-color: var(--palette-lighter-grey);
		-webkit-transition: 0.4s;
		transition: 0.4s;
		border-radius: 2.125rem;
	}

	.slider:before {
		position: absolute;
		content: "";
		height: 1rem;
		width: 1rem;
		left: 0.25rem;
		bottom: 0.25rem;
		background-color: var(--palette-white);
		-webkit-transition: 0.4s;
		transition: 0.4s;
		border-radius: 50%;
	}

	input:checked + .slider {
		background-color: #2196f3;
	}

	input:checked + .slider:before {
		-webkit-transform: translateX(1.625rem);
		-ms-transform: translateX(1.625rem);
		transform: translateX(1.625rem);
	}
</style>
