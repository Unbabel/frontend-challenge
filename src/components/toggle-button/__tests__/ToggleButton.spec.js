import { describe, it, expect, beforeEach, vi } from "vitest";

import { mount } from "@vue/test-utils";

// Components
import ToggleButton from "../ToggleButton.vue";

const mockFn = vi.fn();

describe("ToggleButton", () => {
	let wrapper;
	let testProps = { checked: true, onChangeCallback: mockFn };

	beforeEach(() => {
		// Arrange
		wrapper = mount(ToggleButton, {
			propsData: testProps,
		});
	});

	it("matches snapshot", () => {
		// Assert
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("renders properly", () => {
		// Assert
		const label = wrapper.find("label");
		const input = wrapper.find("input");
		const span = wrapper.find("span");

		expect(label.exists()).toBe(true);
		expect(input.exists()).toBe(true);
		expect(span.exists()).toBe(true);
	});

	it("has correct classes", () => {
		// Assert
		const label = wrapper.find("label");
		const span = wrapper.find("span");

		expect(label.classes()).toContain("switch");
		expect(span.classes()).toContain("slider");
	});

	it("toggles between checked and not checked on click", async () => {
		// Arrange
		const checked = true;
		const notChecked = false;
		const input = await wrapper.find("input");

		// Assert
		expect(input.element.checked).toBe(checked);

		// Act
		await input.trigger("click");

		// Assert
		expect(input.element.checked).toBe(notChecked);
	});
});
