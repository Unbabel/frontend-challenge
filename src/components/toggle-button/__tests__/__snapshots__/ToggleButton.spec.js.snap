// Vitest Snapshot v1

exports[`ToggleButton > matches snapshot 1`] = `"<div data-v-f41a369e=\\"\\"><label data-v-f41a369e=\\"\\" class=\\"switch\\"><input data-v-f41a369e=\\"\\" type=\\"checkbox\\" teste=\\"lol\\"><span data-v-f41a369e=\\"\\" data-test=\\"toggle-button\\" class=\\"slider\\"></span></label></div>"`;
